src/honey_badger.rs

```diff
@@ -200,43 +222,213 @@ where
-        while let Some(ser_batches) = self.take_current_output() {
-            // Deserialize the output.
-            let transactions: BTreeMap<NodeUid, Vec<Tx>> = ser_batches
-                .into_iter()
-                .filter_map(|(proposer_id, ser_batch)| {
-                    // If serialization fails, the proposer of that batch is faulty. Ignore it.
-                    bincode::deserialize::<Vec<Tx>>(&ser_batch)
-                        .ok()
-                        .map(|proposed| (proposer_id, proposed))
-                })
-                .collect();
-            let batch = Batch {
-                epoch: self.epoch,
-                transactions,
-            };
+    /// Handles decryption shares sent by `HoneyBadger` instances.
+    fn handle_decryption_share_message(
+        &mut self,
+        sender_id: &NodeUid,
+        epoch: u64,
+        proposer_id: &NodeUid,
+        share: &DecryptionShare<Bls12>,
+    ) -> HoneyBadgerResult<()> {
+        // Insert the share.
+        self.received_shares
+            .entry(epoch)
+            .and_modify(|epoch_entry| {
+                epoch_entry
+                    .entry(proposer_id.clone())
+                    .and_modify(|proposer_entry| {
+                        proposer_entry.insert(sender_id.clone(), share.clone());
+                    })
+                    .or_insert_with(|| {
+                        vec![(sender_id.clone(), share.clone())]
+                            .into_iter()
+                            .collect()
+                    });
+            })
+            .or_insert_with(|| {
+                vec![(
+                    proposer_id.clone(),
+                    vec![(sender_id.clone(), share.clone())]
+                        .into_iter()
+                        .collect(),
+                )].into_iter()
+                    .collect()
+            });
+
+        if epoch == self.epoch && self.try_decrypt_proposer_selection(proposer_id) {
+            self.try_output_batch()?;
+        }
+
+        Ok(())
+    }
+
+    /// When selections of transactions have been decrypted for all valid proposers in this epoch,
+    /// moves those transactions into a batch, outputs the batch and updates the epoch.
+    fn try_output_batch(&mut self) -> HoneyBadgerResult<bool> {
+        // Wait until selections have been successfully decoded for all proposer nodes with correct
+        // ciphertext outputs.
+        if !self.all_selections_decrypted() {
+            return Ok(false);
+        }
+
+        // Deserialize the output.
+        let transactions: BTreeMap<NodeUid, Vec<Tx>> = self
+            .decrypted_selections
+            .iter()
+            .flat_map(|(proposer_id, ser_batch)| {
+                // If serialization fails, the proposer of that batch is faulty. Ignore it.
+                bincode::deserialize::<Vec<Tx>>(&ser_batch)
+                    .ok()
+                    .map(|proposed| (proposer_id.clone(), proposed))
```

Comment:
Something to keep in mind for #51: We could add another field (`faulty_nodes`?) to `Batch`, to report the nodes whose input didn't deserialize.